README.md
# FusionTrain
Mod for Factorio

This mod adds a fusion train. This doesn't require fuel, but is also relatively expensive.
The train is as fast as the normal train with rocket fuel.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.17.5
Date: 2019-03-30
  Bugfixes:
    - A typo in the changelog has been fixed
---------------------------------------------------------------------------------------------------
Version: 0.17.4
Date: 2019-03-27
  Bugfixes:
    - Fixed a mistake. https://mods.factorio.com/mod/FusionTrain/discussion/5c9c7514bc6a7d000d35189d
---------------------------------------------------------------------------------------------------
Version: 0.17.3
Date: 2019-03-27
  Changes:
    - Improved performance.
---------------------------------------------------------------------------------------------------
Version: 0.17.2
Date: 2019-03-10
  Bugfixes:
    - Fixed typo.
---------------------------------------------------------------------------------------------------
Version: 0.17.1
Date: 2019-03-10
  Major Features:
    - Release 0.17.
  Minor Features:
    - Fuel bar is now filled.
---------------------------------------------------------------------------------------------------
Version: 0.16.2
Date: 2018-07-02
  Changes:
    - Pull request by "shanemadden". Listen for script_raised_built and script_raised_destroy events.
---------------------------------------------------------------------------------------------------
Version: 0.16.1
  Major Features:
    - Release 0.16.
---------------------------------------------------------------------------------------------------
Version: 0.15.1
  Minor Features:
    - Adding setting page.
    - Adding 2 higher levels of fusion trains.
---------------------------------------------------------------------------------------------------
Version: 0.15.0 - First public release
  Major Features:
    - Release 0.15.
    - Adding a fusion train, that doesn't require fuel.

control.lua
require("lib")

function CallRemoteInterface()
	if remote.interfaces["FuelTrainStop"] then
		remote.call("FuelTrainStop", "exclude_from_fuel_schedule", "fusion-locomotive-1")
		remote.call("FuelTrainStop", "exclude_from_fuel_schedule", "fusion-locomotive-2")
		remote.call("FuelTrainStop", "exclude_from_fuel_schedule", "fusion-locomotive-3")
	end
end

function ON_INIT()
	CallRemoteInterface()
end
script.on_init(ON_INIT)

function ON_CONFIGURATION_CHANGED(data)
	CallRemoteInterface()
	
	local mod_name = "FusionTrain"
	if NeedMigration(data,mod_name) then
		local old_version = GetOldVersion(data,mod_name)
		if old_version < "00.17.03" then
			global = {}
		end
	end
end	
script.on_configuration_changed(ON_CONFIGURATION_CHANGED)

function ON_LOAD()
	CallRemoteInterface()
end
script.on_load(ON_LOAD)

data-updates.lua
if not settings.startup['vanilla-order'].value then
	data.raw['item-with-entity-data']['fusion-locomotive-1'].subgroup = "fusion-transport"
	data.raw['item-with-entity-data']['fusion-locomotive-1'].order = "a"
	data.raw['item-with-entity-data']['fusion-locomotive-2'].subgroup = "fusion-transport"
	data.raw['item-with-entity-data']['fusion-locomotive-2'].order = "b"
	data.raw['item-with-entity-data']['fusion-locomotive-3'].subgroup = "fusion-transport"
	data.raw['item-with-entity-data']['fusion-locomotive-3'].order = "c"
end

if settings.startup['fusion-mk-train'].value then
	train_tech()
	train_rec()
end

if mods['bobvehicleequipment'] then
	if settings.startup['bobmods-vehicleequipment-enablevehiclegrids'].value then
		data.raw['locomotive']['fusion-locomotive-1'].equipment_grid = "bob-locomotive"
		if settings.startup['fusion-mk-train'] then
			data.raw['locomotive']['fusion-locomotive-2'].equipment_grid = "bob-locomotive-2"
			data.raw['locomotive']['fusion-locomotive-3'].equipment_grid = "bob-locomotive-3"
		end
	end
end


if not mods['boblogistics'] then
	if settings.startup['fusion-mk-train'].value then
		data.raw['cargo-wagon']['cargo-wagon'].max_speed = 2.4
		data.raw['fluid-wagon']['fluid-wagon'].max_speed = 2.4
	end	
end


if mods['Vehicle Wagon'] and settings.startup['fusion-mk-train'].value then
	for _,wagon in pairs(data.raw['cargo-wagon']) do
		if wagon.name:find("vehicle-wagon",1,true) then	
			wagon.max_speed = 2.4
		end
	end
end

if mods["EvenMoreLight"] then
	for _,train in pairs(data.raw['locomotive']) do	
		if train.name:match("^fusion%-locomotive%-%d$") then
			train.front_light =
				{
					{
						minimum_darkness = 0.3,
						intensity = 0.9,
						size = 60,
					},
					{
						minimum_darkness = 0.3,
						intensity = 0.9,
						size = 60,
					}
				}
			train.stand_by_light =
				{
					{
						minimum_darkness = 0.3,
						color = {b=1},
						shift = {-0.6, -3.5},
						size = 2,
						intensity = 0.5
					},
					{
						minimum_darkness = 0.3,
						color = {b=1},
						shift = {0.6, -3.5},
						size = 2,
						intensity = 0.5
					},
					{
					  minimum_darkness = 0.3,
					  intensity = 0.9,
					  size = 60,
					},
					{
					  minimum_darkness = 0.3,
					  intensity = 0.9,
					  size = 60,
					}
				}
		end				
	end
end

prototypes/FusionTrain.lua
local f_loc = table.deepcopy(data.raw['locomotive']['locomotive'])
f_loc.name = "fusion-locomotive-mk1"
f_loc.minable.result = "fusion-locomotive-mk1"
f_loc.max_speed = 1.2 * 1.15
f_loc.max_power = 600 *1.8 .. "kW"
f_loc.burner.fuel_inventory_size = 0
				
local f_loc2 = table.deepcopy(data.raw['locomotive']['locomotive'])
f_loc2.name = "fusion-locomotive-mk2"
f_loc2.minable.result = "fusion-locomotive-mk2"
f_loc2.max_health = 1350
f_loc2.max_speed = 1.8 * 1.15
f_loc2.max_power = 900 * 1.8 .. "kW"
f_loc2.burner.fuel_inventory_size = 0
f_loc2.reversing_power_modifier = 0.8
f_loc2.braking_force = 15
f_loc2.friction_force = 0.375
f_loc2.air_resistance = 0.005625
				
local f_loc3 = table.deepcopy(data.raw['locomotive']['locomotive'])
f_loc3.name = "fusion-locomotive-mk3"
f_loc3.minable.result = "fusion-locomotive-mk3"
f_loc3.max_health = 1700
f_loc3.max_speed = 2.4 * 1.15
f_loc3.max_power = 1200 * 1.8 .. "kW"
f_loc3.burner.fuel_inventory_size = 0
f_loc3.reversing_power_modifier = 1
f_loc3.braking_force = 20
f_loc3.friction_force = 0.25
f_loc3.air_resistance = 0.00375

				
data:extend({f_loc,f_loc2,f_loc3})


local floc = table.deepcopy(data.raw['item-with-entity-data']['locomotive'])
floc.name = "fusion-locomotive-mk1"
floc.order = "a[train-system]-f[diesel-locomotive]-a[fusion-train]-a[fusion-locomotive-mk1]"
floc.place_result = "fusion-locomotive-mk1"

local floc2 = table.deepcopy(data.raw['item-with-entity-data']['locomotive'])
floc2.name = "fusion-locomotive-mk2"
floc2.order = "a[train-system]-f[diesel-locomotive]-a[fusion-train]-b[fusion-locomotive-mk2]"
floc2.place_result = "fusion-locomotive-mk2"

local floc3 = table.deepcopy(data.raw['item-with-entity-data']['locomotive'])
floc3.name = "fusion-locomotive-mk3"
floc3.order = "a[train-system]-f[diesel-locomotive]-a[fusion-train]-c[fusion-locomotive-mk3]"
floc3.place_result = "fusion-locomotive-mk3"


data:extend({floc,floc2,floc3})


local eloc = table.deepcopy(data.raw['recipe']['locomotive'])
eloc.name = "fusion-locomotive-mk1"
eloc.ingredients =
	{
		{"locomotive", 1},
		{"nuclear-reactor", 1},
		{"electric-engine-unit", 20},
		{"uranium-fuel-cell", 10}	
	}
eloc.result = "fusion-locomotive-mk1"

local eloc2 = table.deepcopy(data.raw['recipe']['locomotive'])
eloc2.name = "fusion-locomotive-mk2"
eloc2.ingredients =
	{
		{"fusion-locomotive-mk1", 1},
		{"nuclear-reactor", 1},
		{"electric-engine-unit", 20},
		{"uranium-fuel-cell", 10}
	}
eloc2.result = "fusion-locomotive-mk2"

local eloc3 = table.deepcopy(data.raw['recipe']['locomotive'])
eloc3.name = "fusion-locomotive-mk3"
eloc3.ingredients =
	{
		{"fusion-locomotive-mk2", 1},
		{"nuclear-reactor", 1},
		{"electric-engine-unit", 20},
		{"uranium-fuel-cell", 10}
	}
eloc3.result = "fusion-locomotive-mk3"


data:extend({eloc})
function train_rec()
	data:extend({eloc2,eloc3})
end


data:extend(
{
	{
		type = "technology",
		name = "fusion-railway",
		icon = "__FusionTrain__/graphics/tech.png",
		icon_size = 128,
		effects =
		{
			{
				type = "unlock-recipe",
				recipe = "fusion-locomotive-mk1"
			}		
		},
		prerequisites = {"railway", "electric-engine", "nuclear-power"},
		unit =
		{
			count = 200,
			ingredients =
			{
				{"science-pack-1", 1},
				{"science-pack-2", 1},
				{"science-pack-3", 1},
				{"production-science-pack", 1},
				{"high-tech-science-pack", 1}
			},
			time = 20
		},
		order = "c-g-a-a-a-a",
	},
})

function train_tech()
	data:extend(
	{
		{
			type = "technology",
			name = "fusion-railway-mk2",
			icon = "__FusionTrain__/graphics/tech-mk2.png",
			icon_size = 128,
			effects =
			{
				{
					type = "unlock-recipe",
					recipe = "fusion-locomotive-mk2"
				}		
			},
			prerequisites = {"fusion-railway"},
			unit =
			{
				count = 300,
				ingredients =
				{
					{"science-pack-1", 1},
					{"science-pack-2", 1},
					{"science-pack-3", 1},
					{"production-science-pack", 1},
					{"high-tech-science-pack", 1}
				},
				time = 30
			},
			order = "c-g-a-a-a-b",
		},
		{
			type = "technology",
			name = "fusion-railway-mk3",
			icon = "__FusionTrain__/graphics/tech-mk3.png",
			icon_size = 128,
			effects =
			{
				{
					type = "unlock-recipe",
					recipe = "fusion-locomotive-mk3"
				}		
			},
			prerequisites = {"fusion-railway-mk2"},
			unit =
			{
				count = 400,
				ingredients =
				{
					{"science-pack-1", 1},
					{"science-pack-2", 1},
					{"science-pack-3", 1},
					{"production-science-pack", 1},
					{"high-tech-science-pack", 1}
				},
				time = 40
			},
			order = "c-g-a-a-a-c",
		},
	})
end

data:extend(
{
	{
		type = "item-subgroup",
		name = "fusion-transport",
		group = "logistics",
		order = "ef",
	},

})









data.lua
local locomotive_1_entity = table.deepcopy(data.raw['locomotive']['locomotive'])
locomotive_1_entity.name = "fusion-locomotive-1"
locomotive_1_entity.minable.result = "fusion-locomotive-1"
locomotive_1_entity.burner = nil
locomotive_1_entity.energy_source = {type = "void"}
				
local locomotive_2_entity = table.deepcopy(data.raw['locomotive']['locomotive'])
locomotive_2_entity.name = "fusion-locomotive-2"
locomotive_2_entity.minable.result = "fusion-locomotive-2"
locomotive_2_entity.max_health = 1350
locomotive_2_entity.max_speed = 1.8 
locomotive_2_entity.max_power = "900kW"
locomotive_2_entity.reversing_power_modifier = 0.8
locomotive_2_entity.braking_force = 15
locomotive_2_entity.friction_force = 0.375
locomotive_2_entity.air_resistance = 0.005625
locomotive_2_entity.burner = nil
locomotive_2_entity.energy_source = {type = "void"}
				
local locomotive_3_entity = table.deepcopy(data.raw['locomotive']['locomotive'])
locomotive_3_entity.name = "fusion-locomotive-3"
locomotive_3_entity.minable.result = "fusion-locomotive-3"
locomotive_3_entity.max_health = 1700
locomotive_3_entity.max_speed = 2.4
locomotive_3_entity.max_power = "1200kW"
locomotive_3_entity.reversing_power_modifier = 1
locomotive_3_entity.braking_force = 20
locomotive_3_entity.friction_force = 0.25
locomotive_3_entity.air_resistance = 0.00375
locomotive_3_entity.burner = nil
locomotive_3_entity.energy_source = {type = "void"}

				
data:extend({locomotive_1_entity,locomotive_2_entity,locomotive_3_entity})


local locomotive_1_item = table.deepcopy(data.raw['item-with-entity-data']['locomotive'])
locomotive_1_item.name = "fusion-locomotive-1"
locomotive_1_item.order = "a[train-system]-f[diesel-locomotive]-a[fusion-train]-a[fusion-locomotive-1]"
locomotive_1_item.place_result = "fusion-locomotive-1"

local locomotive_2_item = table.deepcopy(data.raw['item-with-entity-data']['locomotive'])
locomotive_2_item.name = "fusion-locomotive-2"
locomotive_2_item.order = "a[train-system]-f[diesel-locomotive]-a[fusion-train]-b[fusion-locomotive-2]"
locomotive_2_item.place_result = "fusion-locomotive-2"

local locomotive_3_item = table.deepcopy(data.raw['item-with-entity-data']['locomotive'])
locomotive_3_item.name = "fusion-locomotive-3"
locomotive_3_item.order = "a[train-system]-f[diesel-locomotive]-a[fusion-train]-c[fusion-locomotive-3]"
locomotive_3_item.place_result = "fusion-locomotive-3"


data:extend({locomotive_1_item,locomotive_2_item,locomotive_3_item})


local locomotive_1_recipe = table.deepcopy(data.raw['recipe']['locomotive'])
locomotive_1_recipe.name = "fusion-locomotive-1"
locomotive_1_recipe.ingredients =
	{
		{"locomotive", 1},
		{"nuclear-reactor", 1},
		{"electric-engine-unit", 20},
		{"used-up-uranium-fuel-cell", 20}	
	}
locomotive_1_recipe.result = "fusion-locomotive-1"

local locomotive_2_recipe = table.deepcopy(data.raw['recipe']['locomotive'])
locomotive_2_recipe.name = "fusion-locomotive-2"
locomotive_2_recipe.ingredients =
	{
		{"fusion-locomotive-1", 1},
		{"nuclear-reactor", 1},
		{"electric-engine-unit", 20},
		{"used-up-uranium-fuel-cell", 20}
	}
locomotive_2_recipe.result = "fusion-locomotive-2"

local locomotive_3_recipe = table.deepcopy(data.raw['recipe']['locomotive'])
locomotive_3_recipe.name = "fusion-locomotive-3"
locomotive_3_recipe.ingredients =
	{
		{"fusion-locomotive-2", 1},
		{"nuclear-reactor", 1},
		{"electric-engine-unit", 20},
		{"used-up-uranium-fuel-cell", 20}
	}
locomotive_3_recipe.result = "fusion-locomotive-3"


data:extend({locomotive_1_recipe})
function train_rec()
	data:extend({locomotive_2_recipe,locomotive_3_recipe})
end


data:extend(
{
	{
		type = "technology",
		name = "fusion-railway",
		icon = "__FusionTrain__/graphics/tech.png",
		icon_size = 128,
		effects =
		{
			{
				type = "unlock-recipe",
				recipe = "fusion-locomotive-1"
			}		
		},
		prerequisites = {"production-science-pack", "utility-science-pack", "nuclear-power"},
		unit =
		{
			count = 200,
			ingredients =
			{
				{"automation-science-pack", 4},
				{"logistic-science-pack", 3},
				{"chemical-science-pack", 2},
				{"production-science-pack", 1},
				{"utility-science-pack", 1}
			},
			time = 20
		},
		order = "c-g-a-a-a-a",
	},
})

function train_tech()
	data:extend(
	{
		{
			type = "technology",
			name = "fusion-railway-2",
			icon = "__FusionTrain__/graphics/tech.png",
			icon_size = 128,
			effects =
			{
				{
					type = "unlock-recipe",
					recipe = "fusion-locomotive-2"
				}		
			},
			prerequisites = {"fusion-railway"},
			unit =
			{
				count = 300,
				ingredients =
				{
					{"automation-science-pack", 4},
					{"logistic-science-pack", 3},
					{"chemical-science-pack", 2},
					{"production-science-pack", 1},
					{"utility-science-pack", 1}
				},
				time = 30
			},
			order = "c-g-a-a-a-b",
		},
		{
			type = "technology",
			name = "fusion-railway-3",
			icon = "__FusionTrain__/graphics/tech.png",
			icon_size = 128,
			effects =
			{
				{
					type = "unlock-recipe",
					recipe = "fusion-locomotive-3"
				}		
			},
			prerequisites = {"fusion-railway-2"},
			unit =
			{
				count = 400,
				ingredients =
				{
					{"automation-science-pack", 4},
					{"logistic-science-pack", 3},
					{"chemical-science-pack", 2},
					{"production-science-pack", 1},
					{"utility-science-pack", 1}
				},
				time = 40
			},
			order = "c-g-a-a-a-c",
		},
	})
end

data:extend(
{
	{
		type = "item-subgroup",
		name = "fusion-transport",
		group = "logistics",
		order = "ef",
	},

})